source/versioning.rst

```diff
@@ -0,0 +1,94 @@
+specification.
+
+- **major version**: The major version number advances when a
+  backward incompatible change is made.  For example, this would
+  happen when an existing property in a schema changes meaning.
+  (An exception to this is that when the major version is 0, there
+  are no guarantees of backward compatibility.)
+
+- **minor version**: The minor version number advances when a
+  backward compatible change is made.  For example, this would
+  happen when new properties are added to a schema.
+
+- **patch version**: The patch version number advances when a minor
+  change is made that does not directly affect the file format itself.
+  For example, this would happen when a misspelling or grammatical
+  error in the text is made.
+
+- **pre-release version**: An optional fourth part may also be present
+  following a hyphen to indicate a pre-release version in development.
+  For example, the pre-release of version ``1.2.3`` would be
+  ``1.2.3-dev+a2c4``.
+
+The version of the standard is incremented based on the the change of
+version in the file format or the schemas.  For example, if any of the
+schemas introduced a backward-incompatible change in a particular
+release of the standard, the major number of the standard's version
+would advance.
+
+Handling version mismatches
+---------------------------
+
+Given these conventions, the ASDF standard recommends certain behavior
+of ASDF libraries.  ASDF libraries should endeavour to support as many
+existing versions of the file format and schemas as possible, and use
+the version numbers in the file to act accordingly.  To help libraries
+support older ASDF files, each ASDF standard release includes all
+previously-released schemas (except those that differ only in the
+patch version number).
+
+For future proofing, the library should gracefully handle version
+numbers that are greater than those understood by the library.
+
+- When encountering a **major version** that is greater than the
+  understood version, by default, an exception should be raised.  This
+  behavior may be overridden through explicit user interaction, in
+  which case the library will attempt to handle the element using the
+  conventions of the most recent understood version.
+
+- When encountering a **minor version** that is greater than the
+  understood version, a warning should be emitted, and the library
+  should attempt to handle the element using the conventions of the
+  most recent understood version.
+
+- When encountering a **patch version** that is greater than the
+  understood version, silently ignore the difference and handle the
+  element using the conventions of the most recent understood version.
+
+When writing ASDF files, it is recommended that libraries provide both
+of the following modes of operation:
+
+- Upgrade the file to the latest versions of the file format and
+  schemas understood by the library.
+
+- Preserve the versions of the file format and schemas of the input
+  file.
```

Comment:
> Maybe this option should only be available (in fact I would argue required) for in-place updates to an existing file (as opposed to reading a file into native data structures, then writing out to an entirely new file).

The present problem -- which is resolvable -- is that `pyasdf` always reads the file into native data structures.

> One option would be to allow specifying all the relevant versions when writing a new ASDF file. This would allow dumping a list of versions from an existing file and storing that off in a variable for later use. 

This seems more tenable.  It also would allow for saying "save to ASDF X.Y.Z", which would then look up in the version index which versions of the schemas were the latest shipped with that specific ASDF version and use that.  That seems like a pretty common use case.

This is perhaps an argument for storing the ASDF spec version explicitly in the file.  (In the first bullet point about the "types of versions" above, I say it isn't).  Let's say the library was operating in "version preserving" mode: If you don't include the ASDF spec version in the file, a library would have to "guess" based on the versions of schemas in the file which version to write back out as.  If you include it, it's trivial -- you can use the ASDF spec version to look up in the version index what versions of everything else to use.  So maybe we modify the header of the ASDF file to be:

```
#ASDF A.B.C X.Y.Z
```

where `A.B.C` is the specification version and `X.Y.Z` is the file format version.